asteroid/masknn/attention.py

```diff
@@ -0,0 +1,229 @@
+import torch.nn.functional as F
+import torch.nn as nn
+from torch.nn.modules.activation import MultiheadAttention
+from asteroid.masknn import activations, norms
+import torch
+from asteroid.utils import has_arg
+
+class ImprovedTransformedLayer(nn.Module):
+    def __init__(self, d_model, n_heads, dim_ff, dropout=0., activation="relu", bidirectional=True, norm="gLN"):
+        super(ImprovedTransformedLayer, self).__init__()
+
+        self.mha = MultiheadAttention(d_model, n_heads, dropout=dropout)
+        self.recurrent = nn.LSTM(d_model, dim_ff, bidirectional=bidirectional)
+        self.dropout = nn.Dropout(dropout)
+        ff_inner_dim = 2*dim_ff if bidirectional else dim_ff
+        self.linear = nn.Linear(ff_inner_dim, d_model)
+        self.activation = activations.get(activation)()
+        self.norm_mha = norms.get(norm)(d_model)
+        self.norm_ff = norms.get(norm)(d_model)
+
+    def forward(self, x):
+        x = x.transpose(1, -1)
+        # x is batch, seq_len, channels
+        # self-attention is applied
+        out = self.mha(x, x, x)[0]
+        x = self.dropout(out) + x
+        x = self.norm_mha(x.transpose(1, -1)).transpose(1, -1)
+
+        # lstm is applied
+        out = self.linear(self.dropout(self.activation(self.recurrent(x)[0])))
+        x = self.dropout(out) + x
+        return self.norm_ff(x.transpose(1, -1))
+
+
+class Oladd(nn.Module):
+
+    def __init__(self, chunk_size, hop_size):
+        super(Oladd, self).__init__()
+        self.chunk_size = chunk_size
+        self.hop_size = hop_size
+        self.n_orig_frames = None
+
+    def unfold(self, x):
+        # x is (batch, chan, frames)
+        batch, chan, frames = x.size()
+        assert x.ndim == 3
+        self.n_orig_frames = x.shape[-1]
```

Comment:
I love this caching idea.
We'll need a warning somewhere in the docs that it keeps track of the last input seen, so that you cannot call `unfold` twice on different inputs and expect `fold` to work properly. 
We can add an optional arg on the `fold` that specifies n_orig_frames if we don't want to use the default.

Reply:
Or make it explicit by giving them descriptive names fold_xyz and unfold_xyz (can’t come up with a good name at the moment sorry) and have “non magic” versions h that are simply called fold/unfold

Reply:
why you can't unfold twice ? you can and as long as you fold it two times it is going to work

Reply:
```python
u = unfold(x)
v = unfold(y)
new_x = fold(u)
```
`fold(u)` will use the `self.n_orig_frames` from `y`, not from `x`. So the output size of `new_x` won't be the one of `x` (if y and x have different length). 
Right?

Reply:
Ok i see now your point. we can either enforce to pass the original sequence length to unfold or have a check to prevent double calling. What would be better ?

Reply:
also can have a fully functional version... IDK really